### PR TITLE
Made workshop.start! idempotent, and validated in the unit test

### PR DESCRIPTION
**Pull request by aoby**

### CODE REVIEW

**Comment by ashercodeorg** [comment]
Should `end!` also be idempotent?

**Comment by aoby** [comment]
`end!` already is, as it's not expected to return anything. I added a check in the test to verify that calling `end!` again succeeds and leaves the end time alone.

File: dashboard/test/models/pd/workshop_test.rb [hunk 1]
```diff
@@ -94,20 +94,34 @@ class Pd::WorkshopTest < ActiveSupport::TestCase
...
+    assert_equal started_at, @workshop.reload.started_at
 
     @workshop.end!
     @workshop.reload
     assert_equal 'Ended', @workshop.state
+    ended_at = @workshop.ended_at
+
+    # End should be idempotent. Calling again leaves the original end time.
+    @workshop.end!
+    assert_equal ended_at, @workshop.reload.ended_at
```
Comment (ashercodeorg) [inline dashboard/test/models/pd/workshop_test.rb:133]: This should be broken into multiple tests. For example, I would expect a 'start is idempotent' test that isolates this aspect of the method.

**Comment by ashercodeorg** [comment]
Whether or not callers expected `end` to return something, it did return something (the return of `self.save!` or `nil`). So as to prevent another error of this type, please document that (a) the return value has no meaning or (b) the meaning of the return type (if meaningful).

**Comment by ashercodeorg** [comment]
LGTM.

**Pull request merged**